### PR TITLE
Added BackendDefaults to Virtual Node

### PR DESCRIPTION
**Pull request by sshver**
----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts [hunk 1]
```diff
@@ -97,8 +102,50 @@ export interface VirtualNodeBaseProps {
...
+   * @default - none
+   */
+  readonly certificateType: string;
+
+  /**
+   * Certificate File path or Certificate ARN.
+   *
+   * @default - none
+   */
+  readonly certificate: string[];
```
Comment (dfezzie) [inline packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:144]: These should not be top level properties. These are specific to the client configuration with regard to TLS. There may be other types of backend defaults we could apply in the future.

You could model this more like

```typescript
interface BackendDefaults {
  readonly tlsClientPolicy: TLSClientPolicy;
}

interface TLSClientPolicy {
  /**
   * @default - True
   */
  readonly enforce?: boolean;

  ...

  readonly validation: TLSClientValidation;
}
```

`TLSClientValidation` could be modeled as a `bind` class to enforce the trust is one of file, acmpca, or sds (in the future). We may want to accept the Certificate Authority construct from ACMPCA, which is not yet built. Adam can help with that, we'll just need to coordinate with him

We also want to be aware of the changes that are going to be made with https://github.com/aws/aws-app-mesh-roadmap/issues/34 and be able to accommodate them easily

File: packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts [hunk 2]
```diff
@@ -97,8 +102,50 @@ export interface VirtualNodeBaseProps {
...
+}
+
+/**
+ * Default Configuration for Virtual Nodes
+ */
+export interface BackendDefaults {
+  /**
+   * TLS enforced if True.
+   *
+   * @default - none
```
Comment (dfezzie) [inline packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:121]: The default in our API is true, let's mirror that here

**Review by dfezzie — CHANGES_REQUESTED** [review]

File: packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts [hunk 3]
```diff
@@ -97,8 +102,50 @@ export interface VirtualNodeBaseProps {
...
+  /**
+   * Default Configuration Virtual Node uses to communicate with Vritual Service
+   *
+   * @default - No Config
+   */
+  readonly backendDefaults?: BackendDefaults;
+}
+
+/**
+ * Default Configuration for Virtual Nodes
```
[the remark this answers is not in the document]
Comment (sshver) [reply packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts]: Added descriptive comment.

File: packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts (hunk 1 above)
Comment (sshver) [reply packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts]: updated the model.

File: packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts (hunk 2 above)
Comment (sshver) [reply packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts]: set enforce to true by default.

File: packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts [hunk 4]
```diff
@@ -113,39 +113,135 @@ export interface VirtualNodeBaseProps {
...
+  /**
+   * Client policy for TLS
+   */
+  readonly tlsClientPolicy: TLSClientPolicyProps;
+}
+
+/**
+ * Properties with respect to TLS backend default.
+ */
+export interface TLSClientPolicyProps {
```
Comment (dfezzie) [inline packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:129]: So the `Props` name is reserved for class constructors. (Found this out in a PR I have out now)

Let's have this named `TlsClientPolicyOptions`

File: packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts [hunk 5]
```diff
@@ -113,39 +113,135 @@ export interface VirtualNodeBaseProps {
...
   /**
    * TLS enforced if True.
```
Comment (dfezzie) [inline packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:131]: Maybe we can reword this to be more descriptive

> TLS Connections with downstream server will always be enforced if `True`

Something to that effect.

File: packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts [hunk 6]
```diff
@@ -113,39 +113,135 @@ export interface VirtualNodeBaseProps {
...
    * TLS enforced on these ports. If not specified it is enforced on all ports.
    *
    * @default - none
    */
-  readonly ports?: number[]
+  readonly ports?: number[];
 
   /**
-   * Certificate discovery method, ACM-PCA or Local file hosting.
+   * To enforce the trust is one of file, acmpca, or sds.
```
Comment (dfezzie) [inline packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:145]: Since we haven't introduced SDS yet, let's remove the reference here

We can also reword this to

> Policy used to determine if the TLS certificate the server presents is accepted

File: packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts [hunk 7]
```diff
@@ -113,39 +113,135 @@ export interface VirtualNodeBaseProps {
...
    *
    * @default - none
    */
-  readonly certificateType: string;
+  readonly validation: TLSClientValidation;
```
Comment (dfezzie) [inline packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:149]: Needs to be optional

File: packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts [hunk 8]
```diff
@@ -113,39 +113,135 @@ export interface VirtualNodeBaseProps {
...
+ * Defines the TLS validation context trust.
+ */
+export abstract class TLSClientValidation {
   /**
-   * Certificate File path or Certificate ARN.
-   *
-   * @default - none
+   * TLS validation context trust for a local file
+   */
+  public static fileTrustValidation(props: FileTrustProps): TLSClientValidation {
```
Comment (dfezzie) [inline packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:159]: We could shorten this to `fileTrust(...)`

File: packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts [hunk 9]
```diff
@@ -113,39 +113,135 @@ export interface VirtualNodeBaseProps {
...
+    return new FileTrust(props);
+  }
+
+  /**
+   * TLS validation context trust for AWS Certicate Manager (ACM) certificate.
```
Comment (dfezzie) [inline packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:164]: I want to specify this is acm-pca somehow, but it doesn't flow great :P You have any ideas?

File: packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts [hunk 10]
```diff
@@ -113,39 +113,135 @@ export interface VirtualNodeBaseProps {
...
+}
+
+/**
+ * ACM Trust Properties
+ */
+export interface ACMTrustProps {
+  /**
+   * Amazon Resource Name of the Certificates
+   */
+  readonly certificateAuthorityArns: string[];
```
Comment (dfezzie) [inline packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:183]: this should be an array of `ACM Private Certificate Authorities`. I think I mentioned this in the previous rev, since acm-pca is not modeled in CDK yet this is tricky. Will need to coordinate with Adam on this.

File: packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts [hunk 11]
```diff
@@ -113,39 +113,135 @@ export interface VirtualNodeBaseProps {
...
+   */
+  public abstract bind(scope: cdk.Construct): CfnVirtualNode.TlsValidationContextProperty;
+}
+
+/**
+ * ACM Trust Properties
+ */
+export interface ACMTrustProps {
+  /**
+   * Amazon Resource Name of the Certificates
```
Comment (dfezzie) [inline packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:181]: We should specify ACM PCA here
> Amazon Resource Names (ARN) of trusted ACM Private Certificate Authorities

File: packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts [hunk 12]
```diff
@@ -113,39 +113,135 @@ export interface VirtualNodeBaseProps {
...
+        },
+      },
+    };
+  }
 }
 
+/**
+ * Represents a TLS validation context trust for an AWS Certicate Manager (ACM) certificate.
+ */
+export class ACMTrust extends TLSClientValidation {
```
Comment (dfezzie) [inline packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:224]: This does not need to be exported. Same for above

File: packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts [hunk 13]
```diff
@@ -113,39 +113,135 @@ export interface VirtualNodeBaseProps {
...
+   * TLS validation context trust for AWS Certicate Manager (ACM) certificate.
+   */
+  public static acmTrustValidation(props: ACMTrustProps): TLSClientValidation {
+    return new ACMTrust(props);
+  }
+
+  /**
+   * Returns Trust context based on trust type.
+   */
+  public abstract bind(scope: cdk.Construct): CfnVirtualNode.TlsValidationContextProperty;
```
Comment (dfezzie) [inline packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:173]: Bind should return a `Config` object here

File: packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts (hunk 8 above)
Comment (dfezzie) [reply packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:159]: We could also potentially accept just the path to the certificate chain instead of having a props interface. Makes it simpler to use for customers, but has the drawback of not being as extensible in the future, but there are plenty of ways around that.

File: packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts [hunk 14]
```diff
@@ -189,31 +285,17 @@ abstract class VirtualNodeBase extends cdk.Resource implements IVirtualNode {
...
-            },
-          },
+    // eslint-disable-next-line no-console
+    console.log(backendDefaults.tlsClientPolicy.enforce);
+    this.backendDefaults.push({
+      clientPolicy: {
+        tls: {
+          enforce: backendDefaults.tlsClientPolicy.enforce === undefined ? true : backendDefaults.tlsClientPolicy.enforce,
+          validation: backendDefaults.tlsClientPolicy.validation.bind(this),
+          ports: backendDefaults.tlsClientPolicy.ports ? backendDefaults.tlsClientPolicy.ports : undefined,
```
Comment (dfezzie) [inline packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:295]: This could be simplified to just `backendDefaults.tlsClientPolicy.ports` since it accepts undefined.

File: packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts [hunk 15]
```diff
@@ -189,31 +285,17 @@ abstract class VirtualNodeBase extends cdk.Resource implements IVirtualNode {
...
-                },
-              },
-            },
-          },
+    // eslint-disable-next-line no-console
+    console.log(backendDefaults.tlsClientPolicy.enforce);
+    this.backendDefaults.push({
+      clientPolicy: {
+        tls: {
+          enforce: backendDefaults.tlsClientPolicy.enforce === undefined ? true : backendDefaults.tlsClientPolicy.enforce,
```
Comment (dfezzie) [inline packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:293]: I think `enforce: backendDefaults.tlsClientPolicy.enforce ?? true` would work here?

**Review by dfezzie — CHANGES_REQUESTED** [review]
Nice work! This will set up an easy win for adding the `ClientPolicy` to backends!

File: packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts (hunk 7 above)
Comment (sshver) [reply packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:159]: Shouldn't it be required ? I found this [reference](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-appmesh-virtualnode-clientpolicytls.html#cfn-appmesh-virtualnode-clientpolicytls-validation) doc where in it says this field in `required`

File: packages/@aws-cdk/aws-appmesh/lib/validation-context.ts [hunk 16]
```diff
@@ -0,0 +1,146 @@
+import * as cdk from '@aws-cdk/core';
+import { CfnVirtualNode } from './appmesh.generated';
+
+/**
+ * Defines the TLS validation context trust.
+ */
+export abstract class TLSClientValidation {
+  /**
+   * TLS validation context trust for a local file
```
Comment (dfezzie) [inline packages/@aws-cdk/aws-appmesh/lib/validation-context.ts:9]: nit: `TLS validation provided from the local file system`

This struct tells the envoy where to fetch the validation context from

same comment below

File: packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts [hunk 17]
```diff
@@ -94,6 +95,13 @@ export interface VirtualNodeBaseProps {
    * @default - No access logging
    */
   readonly accessLog?: AccessLog;
+
+  /**
+   * Default Configuration Virtual Node uses to communicate with Vritual Service
```
Comment (dfezzie) [inline packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:100]: typo: `Vritual` => `Virtual`

File: packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts [hunk 18]
```diff
@@ -175,6 +183,7 @@ export class VirtualNode extends VirtualNodeBase {
    */
   public readonly mesh: IMesh;
 
+  private readonly backendDefaults = new Array<CfnVirtualNode.BackendDefaultsProperty>();
```
Comment (dfezzie) [inline packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:186]: Why use an `Array` here? This could just be `CfnVirtualNode.BackendDefaultsProperty`

File: packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts [hunk 19]
```diff
@@ -216,6 +229,24 @@ export class VirtualNode extends VirtualNodeBase {
       resourceName: this.physicalName,
     });
   }
+  /**
+   * Adds Default Backend Configuration for virtual node to communicate with Virtual Services.
+   */
+  public addBackendDefaults(clientPolicy: ClientPolicy) {
```
Comment (dfezzie) [inline packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:235]: Why would we need this method?

File: packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts [hunk 20]
```diff
@@ -216,6 +229,24 @@ export class VirtualNode extends VirtualNodeBase {
...
+    if (this.backendDefaults.length === 0) {
+      this.backendDefaults.push({
+        clientPolicy: {
+          tls: {
+            enforce: clientPolicy.tlsClientPolicy.enforce ?? true,
```
Comment (dfezzie) [inline packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:240]: Technically we could just do `clientPolicy.tlsClientPolicy.enforce` and accept the `undefined` since the API defaults to `true`. Not sure what the right thing to do here for CDK is

**Review by dfezzie — CHANGES_REQUESTED** [review]
Looks good to me. Just a few comments.

Any reason not to implement this for Virtual Gateways?

File: packages/@aws-cdk/aws-appmesh/lib/validation-context.ts [hunk 21]
```diff
@@ -0,0 +1,146 @@
+import * as cdk from '@aws-cdk/core';
+import { CfnVirtualNode } from './appmesh.generated';
+
+/**
+ * Defines the TLS validation context trust.
+ */
+export abstract class TLSClientValidation {
```
Comment (alexbrjo) [inline packages/@aws-cdk/aws-appmesh/lib/validation-context.ts:7]: Backend Client Policies won't be the only shapes that support validation contexts. Listeners could support validation contexts in the future (ex. mTLS).

Unless there is a limitation that forces us to have 2 different shapes for client validation and server validation, I'd recommend changing the name here to a more general `TLSValidationContext`. And creating a separate file for client policy shapes.

File: packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts (hunk 18 above)
Comment (sshver) [reply packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:186]: If the `backendDefaults` is of type `CfnVirtualNode.BackendDefaultsProperty` and a private member of VirtualNode, I would have to initialize it say, with an empty object because it cannot be undefined. There are no options to omit the empty object and all the resources would have `backendDefaults: {}`, if not provided by the client. Whereas if we use `Array` we could set `omitEmptyArray` true.

File: packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts (hunk 19 above)
Comment (sshver) [reply packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:235]: I was under the impression that the backendDefault can be added to virtual node after it has been created. Hence a public method.

File: packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts (hunk 18 above)
Comment (sshver) [reply packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:186]: Or something like this would work as well.

```typescript
    if (props.backendDefaults) {
      this.backendDefaults = this.addBackendDefaults(props.backendDefaults);
    } else {
      this.backendDefaults = {};
    }

 const node = new CfnVirtualGateway(this, 'Resource', {
     backendDefaults: Object.keys(this.backendDefaults).length != 0 ? this.backendDefaults : undefined,
. . .
 });
```
